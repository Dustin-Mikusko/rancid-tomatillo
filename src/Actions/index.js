export const setMovies = movies => ({
  type: 'SET_MOVIES',
  movies
});

export const changeLoading = isLoading => ({
  type: 'CHANGE_LOADING',
  isLoading: !isLoading
});

export const addUser = user => ({
  type: 'ADD_USER',
  user
});
