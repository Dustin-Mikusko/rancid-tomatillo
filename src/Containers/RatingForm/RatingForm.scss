* {
  font-family: 'Quicksand', sans-serif;
}

.rating-form {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: baseline;
  width: 100%;
}

.rating-btn {
  font-family: 'Quicksand', sans-serif;
  background: #000;
  border: 2px ridge #FFF;
  border-radius: 3px;
  font-size: 100%;
  font-weight: 700;
}

select {
  width: 90px;
  height: 30px;
  background-color: #000;
  border: 2px ridge #FFF;
  font-size: 100%;
  font-weight: 700;
}
