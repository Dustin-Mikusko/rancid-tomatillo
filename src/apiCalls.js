export const getMovies = () => {
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/movies')
    .then(res => {
      if (!res.ok) {
        throw Error('Oops! The Box Office must be closed!')
      }
      return res.json();
    })
};

export const getUser = (user) => {
  const options = {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
    .then(res => {
      if (!res.ok) {
        throw Error('Incorrect Email and/or Password')
      }
      return res.json();
    })
};

export const getUserRatings = id => {
  return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${id}/ratings`).then(res => {
    if (!res.ok) {
      throw Error('Could not find your ratings')
    }
    return res.json();
  })
}
