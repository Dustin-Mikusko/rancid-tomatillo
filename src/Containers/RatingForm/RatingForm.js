import React, { Component } from 'react';
import { postRating, getUserRatings } from '../../apiCalls';
import { updateRatings } from '../../Actions';
import { connect } from 'react-redux';
import { getMovies } from '../../apiCalls';
import { setMovies } from '../../Actions';




class RatingForm extends Component {
  constructor() {
    super();
    this.state = {
      currentRating: null
    }
  }

  changeRating = number => {
    this.setState({ currentRating: parseInt(number) })
  }

  render() {
    return (
      <>
        <select name='rating' onChange={e => this.changeRating(e.target.value)}>
          <option value=''>--Rate Movie--</option>
          <option value={1}>1</option>
          <option value={2}>2</option>
          <option value={3}>3</option>
          <option value={4}>4</option>
          <option value={5}>5</option>
          <option value={6}>6</option>
          <option value={7}>7</option>
          <option value={8}>8</option>
          <option value={9}>9</option>
          <option value={10}>10</option>
        </select>
        <button onClick={() => {
          postRating(this.props.movieId, this.state.currentRating, this.props.userId)
            .then(data => {
              getUserRatings(data.rating.user_id)
                .then(ratings => {
                  this.props.updateRatings(ratings.ratings);
                  getMovies()
                  .then(data => {
                    this.props.setMovies(data.movies);
                  })
                })
            })
        }}>SUBMIT.....DAT.....RATING</button>
      </>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  updateRatings: ratings => dispatch( updateRatings(ratings) ),
  setMovies: (movies) => dispatch(setMovies(movies))
})

export default connect(null, mapDispatchToProps)(RatingForm);
