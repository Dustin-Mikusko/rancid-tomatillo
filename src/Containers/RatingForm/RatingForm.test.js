import React from 'react';
import { RatingForm, mapDispatchToProps } from './RatingForm';
import { shallow } from 'enzyme';
import { updateRatings, setMovies } from '../../Actions';
import { postRating, getUserRatings, getMovies, deleteRating } from '../../apiCalls';

jest.mock('../../apiCalls.js');


describe('RatingForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<RatingForm 
      userId={4}
      ratingId={23984}
      movieId={12}
    />)
  });

  it('should match snapshot if there is no userRating', () => {
    const wrapper = shallow(<RatingForm />);

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should match snapshot if there is a userRating', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  describe('changeRating', () => {
    it('should set the local state to the selected value in the form', () => {
      let mockDefaultState = {
        currentRating: null
      };
      let expected = {
        currentRating: 4
      };
      expect(wrapper.state()).toEqual(mockDefaultState);

      wrapper.instance().changeRating(4);

      expect(wrapper.state()).toEqual(expected);
    });

    it('should update local state onChange of select', () => {
      let mockDefaultState = {
        currentRating: null
      };
      const mockEvent = {
        target: {
          value: 2
        }
      }
      expect(wrapper.state()).toEqual(mockDefaultState);

      wrapper.find('select').simulate('change', mockEvent);

      expect(wrapper.state('currentRating')).toEqual(2)
    })
  });

  describe('submitRating', () => {
    let wrapper, mockUpdateRatings, mockSetMovies;
    beforeEach(() => {
      mockUpdateRatings = jest.fn();
      mockSetMovies = jest.fn();
      wrapper = shallow(<RatingForm
        updateRatings={mockUpdateRatings}
        setMovies={mockSetMovies}
        ratingId={undefined}
        movieId={12}
        userId={4}
      />);
      postRating.mockImplementation(() => {
        return Promise.resolve({
          rating: {
            user_id: 4
          }
        })
      });
      getUserRatings.mockImplementation(() => {
        return Promise.resolve({
          ratings: [
            {
              id: 1, 
              user_id: 1, 
              movie_id: 1, 
              rating: 6, 
              created_at: "someDate", 
              updated_at: "someDate"
            }
        ]
      })
      });
      getMovies.mockImplementation(() => {
        return Promise.resolve({
          movies: [
            {
              id: 1, 
              title: "Movie Title",
              poster_path: "someURL", 
              backdrop_path: "someURL", 
              release_date: "2019-12-04", 
              overview: "Some overview", 
              average_rating: 6 }
          ]
        })
      })
    });
    it('should invoke submitRating on button click when there is no userRating', () => {
      wrapper.instance().submitRating = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('#submit-btn').simulate('click');

      expect(wrapper.instance().submitRating).toHaveBeenCalled();
    });

    it('should invoke postRating when submitRating is fired', () => {
      wrapper.instance().changeRating(4);
      wrapper.instance().submitRating();

      expect(postRating).toHaveBeenCalledWith(12, 4, 4)
    });

    it('should invoke getUserRatings when submitRating is fired', () => {
      wrapper.instance().changeRating(4);
      wrapper.instance().submitRating();

      expect(getUserRatings).toHaveBeenCalledWith(4);
    });

    it('should invoke getMovies when submitRating is fired', () => {
      wrapper.instance().submitRating();

      expect(getMovies).toHaveBeenCalled();
    });
  });

  describe('updateRating', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<RatingForm
        ratingId={23984}
        movieId={12}
        userId={4}
      />);
      deleteRating.mockImplementation(() => Promise.resolve({name: 'name'}));
      
    });

    it('should invoke deleteRating and submitRating when updateRating is fired', async () => {
      wrapper.instance().submitRating = jest.fn();
      await wrapper.instance().updateRating();

      expect(deleteRating).toHaveBeenCalledWith(4, 23984);
      expect(wrapper.instance().submitRating).toHaveBeenCalled();
    });

    it('should invoke updateRating on button click when there is a userRating', () => {
      wrapper.instance().updateRating = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('#update-btn').simulate('click');

      expect(wrapper.instance().updateRating).toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with an updateRatings action when submitRating is called', () => {
      let mockDispatch = jest.fn();
      let mockAction = updateRatings('WADDUP');

      let mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateRatings('WADDUP');

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('should call dispatch with a setMovies action when submitRating is called', () => {
      let mockDispatch = jest.fn();
      let mockAction = setMovies('moviesArray');

      let mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setMovies('moviesArray');

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    })
  })
});
