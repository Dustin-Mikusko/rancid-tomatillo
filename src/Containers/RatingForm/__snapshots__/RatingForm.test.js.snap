// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingForm should match snapshot if there is a userRating 1`] = `
"<Fragment>
  <select name=\\"rating\\" onChange={[Function: onChange]}>
    <option value=\\"\\">
      --Rate Movie--
    </option>
    <option value={1}>
      1
    </option>
    <option value={2}>
      2
    </option>
    <option value={3}>
      3
    </option>
    <option value={4}>
      4
    </option>
    <option value={5}>
      5
    </option>
    <option value={6}>
      6
    </option>
    <option value={7}>
      7
    </option>
    <option value={8}>
      8
    </option>
    <option value={9}>
      9
    </option>
    <option value={10}>
      10
    </option>
  </select>
  <button id=\\"update-btn\\" onClick={[Function]}>
    UPDATE.....DAT.....RATING
  </button>
</Fragment>"
`;

exports[`RatingForm should match snapshot if there is no userRating 1`] = `
"<Fragment>
  <select name=\\"rating\\" onChange={[Function: onChange]}>
    <option value=\\"\\">
      --Rate Movie--
    </option>
    <option value={1}>
      1
    </option>
    <option value={2}>
      2
    </option>
    <option value={3}>
      3
    </option>
    <option value={4}>
      4
    </option>
    <option value={5}>
      5
    </option>
    <option value={6}>
      6
    </option>
    <option value={7}>
      7
    </option>
    <option value={8}>
      8
    </option>
    <option value={9}>
      9
    </option>
    <option value={10}>
      10
    </option>
  </select>
  <button id=\\"submit-btn\\" onClick={[Function]}>
    SUBMIT.....DAT.....RATING
  </button>
</Fragment>"
`;
